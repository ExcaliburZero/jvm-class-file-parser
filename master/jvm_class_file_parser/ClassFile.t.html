<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=struct.ClassFile.html">
</head>
<body>
    <p>Redirecting to <a href="struct.ClassFile.html">struct.ClassFile.html</a>...</p>
    <script>location.replace("struct.ClassFile.html" + location.search + location.hash);</script>
</body>
</html>