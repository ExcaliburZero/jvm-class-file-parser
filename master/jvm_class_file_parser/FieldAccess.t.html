<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=enum.FieldAccess.html">
</head>
<body>
    <p>Redirecting to <a href="enum.FieldAccess.html">enum.FieldAccess.html</a>...</p>
    <script>location.replace("enum.FieldAccess.html" + location.search + location.hash);</script>
</body>
</html>