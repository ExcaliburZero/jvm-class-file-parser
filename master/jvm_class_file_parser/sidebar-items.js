initSidebarItems({"enum":[["Bytecode","A JVM bytecode instruction."],["ClassAccess","A flag that denotes an access level or property of a class."],["ConstantPoolEntry",""],["FieldAccess","A flag that denotes an acess level or property of a field."]],"struct":[["Attribute",""],["ClassFile","A representation of a JVM class file."],["Code",""],["ExceptionTableEntry",""],["Field",""],["Method",""]]});