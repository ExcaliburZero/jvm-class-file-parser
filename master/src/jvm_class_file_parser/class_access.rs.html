<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Source to the Rust file `src/class_access.rs`."><meta name="keywords" content="rust, rustlang, rust-lang"><title>class_access.rs.html -- source</title><link rel="stylesheet" type="text/css" href="../../normalize.css"><link rel="stylesheet" type="text/css" href="../../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../../dark.css"><link rel="stylesheet" type="text/css" href="../../light.css" id="themeStyle"><script src="../../storage.js"></script><noscript><link rel="stylesheet" href="../../noscript.css"></noscript><link rel="shortcut icon" href="../../favicon.ico"><style type="text/css">#crate-search{background-image:url("../../down-arrow.svg");}</style></head><body class="rustdoc source"><!--[if lte IE 8]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="sidebar-menu">&#9776;</div><a href='../../jvm_class_file_parser/index.html'><img src='../../rust-logo.png' alt='logo' width='100'></a></nav><div class="theme-picker"><button id="theme-picker" aria-label="Pick another theme!"><img src="../../brush.svg" width="18" alt="Pick another theme!"></button><div id="theme-choices"></div></div><script src="../../theme.js"></script><nav class="sub"><form class="search-form js-only"><div class="search-container"><div><select id="crate-search"><option value="All crates">All crates</option></select><input class="search-input" name="search" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"></div><a id="settings-menu" href="../../settings.html"><img src="../../wheel.svg" width="18" alt="Change settings"></a></div></form></nav><section id="main" class="content"><pre class="line-numbers"><span id="1">  1</span>
<span id="2">  2</span>
<span id="3">  3</span>
<span id="4">  4</span>
<span id="5">  5</span>
<span id="6">  6</span>
<span id="7">  7</span>
<span id="8">  8</span>
<span id="9">  9</span>
<span id="10"> 10</span>
<span id="11"> 11</span>
<span id="12"> 12</span>
<span id="13"> 13</span>
<span id="14"> 14</span>
<span id="15"> 15</span>
<span id="16"> 16</span>
<span id="17"> 17</span>
<span id="18"> 18</span>
<span id="19"> 19</span>
<span id="20"> 20</span>
<span id="21"> 21</span>
<span id="22"> 22</span>
<span id="23"> 23</span>
<span id="24"> 24</span>
<span id="25"> 25</span>
<span id="26"> 26</span>
<span id="27"> 27</span>
<span id="28"> 28</span>
<span id="29"> 29</span>
<span id="30"> 30</span>
<span id="31"> 31</span>
<span id="32"> 32</span>
<span id="33"> 33</span>
<span id="34"> 34</span>
<span id="35"> 35</span>
<span id="36"> 36</span>
<span id="37"> 37</span>
<span id="38"> 38</span>
<span id="39"> 39</span>
<span id="40"> 40</span>
<span id="41"> 41</span>
<span id="42"> 42</span>
<span id="43"> 43</span>
<span id="44"> 44</span>
<span id="45"> 45</span>
<span id="46"> 46</span>
<span id="47"> 47</span>
<span id="48"> 48</span>
<span id="49"> 49</span>
<span id="50"> 50</span>
<span id="51"> 51</span>
<span id="52"> 52</span>
<span id="53"> 53</span>
<span id="54"> 54</span>
<span id="55"> 55</span>
<span id="56"> 56</span>
<span id="57"> 57</span>
<span id="58"> 58</span>
<span id="59"> 59</span>
<span id="60"> 60</span>
<span id="61"> 61</span>
<span id="62"> 62</span>
<span id="63"> 63</span>
<span id="64"> 64</span>
<span id="65"> 65</span>
<span id="66"> 66</span>
<span id="67"> 67</span>
<span id="68"> 68</span>
<span id="69"> 69</span>
<span id="70"> 70</span>
<span id="71"> 71</span>
<span id="72"> 72</span>
<span id="73"> 73</span>
<span id="74"> 74</span>
<span id="75"> 75</span>
<span id="76"> 76</span>
<span id="77"> 77</span>
<span id="78"> 78</span>
<span id="79"> 79</span>
<span id="80"> 80</span>
<span id="81"> 81</span>
<span id="82"> 82</span>
<span id="83"> 83</span>
<span id="84"> 84</span>
<span id="85"> 85</span>
<span id="86"> 86</span>
<span id="87"> 87</span>
<span id="88"> 88</span>
<span id="89"> 89</span>
<span id="90"> 90</span>
<span id="91"> 91</span>
<span id="92"> 92</span>
<span id="93"> 93</span>
<span id="94"> 94</span>
<span id="95"> 95</span>
<span id="96"> 96</span>
<span id="97"> 97</span>
<span id="98"> 98</span>
<span id="99"> 99</span>
<span id="100">100</span>
<span id="101">101</span>
<span id="102">102</span>
<span id="103">103</span>
<span id="104">104</span>
<span id="105">105</span>
<span id="106">106</span>
<span id="107">107</span>
<span id="108">108</span>
<span id="109">109</span>
<span id="110">110</span>
<span id="111">111</span>
<span id="112">112</span>
<span id="113">113</span>
<span id="114">114</span>
</pre><div class="example-wrap"><pre class="rust ">
<span class="kw">use</span> <span class="ident">std</span>::<span class="ident">collections</span>::<span class="ident">HashSet</span>;

<span class="kw">use</span> <span class="ident">util</span>::<span class="ident">flag_is_set</span>;

<span class="comment">// Access flag masks are from Table 4.1-B of the JVM specification</span>
<span class="comment">//</span>
<span class="comment">// https://docs.oracle.com/javase/specs/jvms/se11/html/jvms-4.html#jvms-4.1-200-E.1</span>
<span class="kw">const</span> <span class="ident">PUBLIC_FLAG</span>: <span class="ident">u16</span> <span class="op">=</span> <span class="number">0x0001</span>;
<span class="kw">const</span> <span class="ident">FINAL_FLAG</span>: <span class="ident">u16</span> <span class="op">=</span> <span class="number">0x0010</span>;
<span class="kw">const</span> <span class="ident">SUPER_FLAG</span>: <span class="ident">u16</span> <span class="op">=</span> <span class="number">0x0020</span>;
<span class="kw">const</span> <span class="ident">INTERFACE_FLAG</span>: <span class="ident">u16</span> <span class="op">=</span> <span class="number">0x0200</span>;
<span class="kw">const</span> <span class="ident">ABSTRACT_FLAG</span>: <span class="ident">u16</span> <span class="op">=</span> <span class="number">0x0400</span>;
<span class="kw">const</span> <span class="ident">SYNTHETIC_FLAG</span>: <span class="ident">u16</span> <span class="op">=</span> <span class="number">0x1000</span>;
<span class="kw">const</span> <span class="ident">ANNOTATION_FLAG</span>: <span class="ident">u16</span> <span class="op">=</span> <span class="number">0x2000</span>;
<span class="kw">const</span> <span class="ident">ENUM_FLAG</span>: <span class="ident">u16</span> <span class="op">=</span> <span class="number">0x4000</span>;
<span class="kw">const</span> <span class="ident">MODULE_FLAG</span>: <span class="ident">u16</span> <span class="op">=</span> <span class="number">0x8000</span>;

<span class="doccomment">/// A flag that denotes an access level or property of a class.</span>
<span class="doccomment">///</span>
<span class="doccomment">/// See the `access_flags` section of Chapter 4.1 of the JVM specification for</span>
<span class="doccomment">/// details.</span>
<span class="doccomment">///</span>
<span class="doccomment">/// https://docs.oracle.com/javase/specs/jvms/se11/html/jvms-4.html#jvms-4.1-200-E</span>
<span class="attribute">#[<span class="ident">derive</span>(<span class="ident">Clone</span>)]</span>
<span class="attribute">#[<span class="ident">derive</span>(<span class="ident">Debug</span>)]</span>
<span class="attribute">#[<span class="ident">derive</span>(<span class="ident">Eq</span>)]</span>
<span class="attribute">#[<span class="ident">derive</span>(<span class="ident">Hash</span>)]</span>
<span class="attribute">#[<span class="ident">derive</span>(<span class="ident">Ord</span>)]</span>
<span class="attribute">#[<span class="ident">derive</span>(<span class="ident">PartialEq</span>)]</span>
<span class="attribute">#[<span class="ident">derive</span>(<span class="ident">PartialOrd</span>)]</span>
<span class="kw">pub</span> <span class="kw">enum</span> <span class="ident">ClassAccess</span> {
    <span class="ident">Public</span>,
    <span class="ident">Final</span>,
    <span class="ident">Super</span>,
    <span class="ident">Interface</span>,
    <span class="ident">Abstract</span>,
    <span class="ident">Synthetic</span>,
    <span class="ident">Annotation</span>,
    <span class="ident">Enum</span>,
    <span class="ident">Module</span>,
}

<span class="kw">impl</span> <span class="ident">ClassAccess</span> {
    <span class="doccomment">/// Extracts the list of class access flags that are embedded in the given</span>
    <span class="doccomment">/// access flag value.</span>
    <span class="doccomment">///</span>
    <span class="doccomment">/// Returns an error message if the extracted combination of access flags</span>
    <span class="doccomment">/// are inconsistent. (This validation has not yet been implemented)</span>
    <span class="doccomment">///</span>
    <span class="doccomment">/// See Table 4.1-B of the JVM specification for more details.</span>
    <span class="doccomment">///</span>
    <span class="doccomment">/// https://docs.oracle.com/javase/specs/jvms/se11/html/jvms-4.html#jvms-4.1-200-E.1</span>
    <span class="doccomment">///</span>
    <span class="doccomment">/// ```</span>
    <span class="doccomment">/// # use std::collections::HashSet;</span>
    <span class="doccomment">/// # use jvm_class_file_parser::ClassAccess;</span>
    <span class="doccomment">/// #</span>
    <span class="doccomment">/// let access_flags = 0b0000_0010_0000_0001;</span>
    <span class="doccomment">///</span>
    <span class="doccomment">/// let mut expected = HashSet::new();</span>
    <span class="doccomment">/// expected.insert(ClassAccess::Public);</span>
    <span class="doccomment">/// expected.insert(ClassAccess::Interface);</span>
    <span class="doccomment">///</span>
    <span class="doccomment">/// assert_eq!(Ok(expected), ClassAccess::from_access_flags(access_flags));</span>
    <span class="doccomment">/// ```</span>
    <span class="kw">pub</span> <span class="kw">fn</span> <span class="ident">from_access_flags</span>(<span class="ident">access_flags</span>: <span class="ident">u16</span>) <span class="op">-&gt;</span> <span class="prelude-ty">Result</span><span class="op">&lt;</span><span class="ident">HashSet</span><span class="op">&lt;</span><span class="ident">ClassAccess</span><span class="op">&gt;</span>, <span class="ident">String</span><span class="op">&gt;</span> {
        <span class="kw">use</span> <span class="ident">ClassAccess</span>::<span class="kw-2">*</span>;

        <span class="kw">let</span> <span class="kw-2">mut</span> <span class="ident">access</span> <span class="op">=</span> <span class="ident">HashSet</span>::<span class="ident">new</span>();

        <span class="kw">let</span> <span class="ident">is_public</span> <span class="op">=</span> <span class="ident">flag_is_set</span>(<span class="ident">PUBLIC_FLAG</span>, <span class="ident">access_flags</span>);
        <span class="kw">let</span> <span class="ident">is_final</span> <span class="op">=</span> <span class="ident">flag_is_set</span>(<span class="ident">FINAL_FLAG</span>, <span class="ident">access_flags</span>);
        <span class="kw">let</span> <span class="ident">is_super</span> <span class="op">=</span> <span class="ident">flag_is_set</span>(<span class="ident">SUPER_FLAG</span>, <span class="ident">access_flags</span>);
        <span class="kw">let</span> <span class="ident">is_interface</span> <span class="op">=</span> <span class="ident">flag_is_set</span>(<span class="ident">INTERFACE_FLAG</span>, <span class="ident">access_flags</span>);
        <span class="kw">let</span> <span class="ident">is_abstract</span> <span class="op">=</span> <span class="ident">flag_is_set</span>(<span class="ident">ABSTRACT_FLAG</span>, <span class="ident">access_flags</span>);
        <span class="kw">let</span> <span class="ident">is_synthetic</span> <span class="op">=</span> <span class="ident">flag_is_set</span>(<span class="ident">SYNTHETIC_FLAG</span>, <span class="ident">access_flags</span>);
        <span class="kw">let</span> <span class="ident">is_annotation</span> <span class="op">=</span> <span class="ident">flag_is_set</span>(<span class="ident">ANNOTATION_FLAG</span>, <span class="ident">access_flags</span>);
        <span class="kw">let</span> <span class="ident">is_enum</span> <span class="op">=</span> <span class="ident">flag_is_set</span>(<span class="ident">ENUM_FLAG</span>, <span class="ident">access_flags</span>);
        <span class="kw">let</span> <span class="ident">is_module</span> <span class="op">=</span> <span class="ident">flag_is_set</span>(<span class="ident">MODULE_FLAG</span>, <span class="ident">access_flags</span>);

        <span class="comment">// TODO: Add validation for inconsistent access flags</span>

        <span class="kw">if</span> <span class="ident">is_public</span> { <span class="ident">access</span>.<span class="ident">insert</span>(<span class="ident">Public</span>); }
        <span class="kw">if</span> <span class="ident">is_final</span> { <span class="ident">access</span>.<span class="ident">insert</span>(<span class="ident">Final</span>); }
        <span class="kw">if</span> <span class="ident">is_super</span> { <span class="ident">access</span>.<span class="ident">insert</span>(<span class="ident">Super</span>); }
        <span class="kw">if</span> <span class="ident">is_interface</span> { <span class="ident">access</span>.<span class="ident">insert</span>(<span class="ident">Interface</span>); }
        <span class="kw">if</span> <span class="ident">is_abstract</span> { <span class="ident">access</span>.<span class="ident">insert</span>(<span class="ident">Abstract</span>); }
        <span class="kw">if</span> <span class="ident">is_synthetic</span> { <span class="ident">access</span>.<span class="ident">insert</span>(<span class="ident">Synthetic</span>); }
        <span class="kw">if</span> <span class="ident">is_annotation</span> { <span class="ident">access</span>.<span class="ident">insert</span>(<span class="ident">Annotation</span>); }
        <span class="kw">if</span> <span class="ident">is_enum</span> { <span class="ident">access</span>.<span class="ident">insert</span>(<span class="ident">Enum</span>); }
        <span class="kw">if</span> <span class="ident">is_module</span> { <span class="ident">access</span>.<span class="ident">insert</span>(<span class="ident">Module</span>); }

        <span class="prelude-val">Ok</span>(<span class="ident">access</span>)
    }

    <span class="attribute">#[<span class="ident">allow</span>(<span class="ident">clippy</span>::<span class="ident">wrong_self_convention</span>)]</span>
    <span class="kw">pub</span> <span class="kw">fn</span> <span class="ident">to_access_flags</span>(<span class="ident">access_flags</span>: <span class="kw-2">&amp;</span><span class="ident">HashSet</span><span class="op">&lt;</span><span class="ident">ClassAccess</span><span class="op">&gt;</span>) <span class="op">-&gt;</span> <span class="ident">u16</span> {
        <span class="kw">use</span> <span class="ident">ClassAccess</span>::<span class="kw-2">*</span>;

        <span class="kw">let</span> <span class="kw-2">mut</span> <span class="ident">flags</span> <span class="op">=</span> <span class="number">0</span>;

        <span class="kw">if</span> <span class="ident">access_flags</span>.<span class="ident">contains</span>(<span class="kw-2">&amp;</span><span class="ident">Public</span>) { <span class="ident">flags</span> <span class="op">^=</span> <span class="ident">PUBLIC_FLAG</span>; }
        <span class="kw">if</span> <span class="ident">access_flags</span>.<span class="ident">contains</span>(<span class="kw-2">&amp;</span><span class="ident">Final</span>) { <span class="ident">flags</span> <span class="op">^=</span> <span class="ident">FINAL_FLAG</span>; }
        <span class="kw">if</span> <span class="ident">access_flags</span>.<span class="ident">contains</span>(<span class="kw-2">&amp;</span><span class="ident">Super</span>) { <span class="ident">flags</span> <span class="op">^=</span> <span class="ident">SUPER_FLAG</span>; }
        <span class="kw">if</span> <span class="ident">access_flags</span>.<span class="ident">contains</span>(<span class="kw-2">&amp;</span><span class="ident">Interface</span>) { <span class="ident">flags</span> <span class="op">^=</span> <span class="ident">INTERFACE_FLAG</span>; }
        <span class="kw">if</span> <span class="ident">access_flags</span>.<span class="ident">contains</span>(<span class="kw-2">&amp;</span><span class="ident">Abstract</span>) { <span class="ident">flags</span> <span class="op">^=</span> <span class="ident">ABSTRACT_FLAG</span>; }
        <span class="kw">if</span> <span class="ident">access_flags</span>.<span class="ident">contains</span>(<span class="kw-2">&amp;</span><span class="ident">Synthetic</span>) { <span class="ident">flags</span> <span class="op">^=</span> <span class="ident">SYNTHETIC_FLAG</span>; }
        <span class="kw">if</span> <span class="ident">access_flags</span>.<span class="ident">contains</span>(<span class="kw-2">&amp;</span><span class="ident">Annotation</span>) { <span class="ident">flags</span> <span class="op">^=</span> <span class="ident">ANNOTATION_FLAG</span>; }
        <span class="kw">if</span> <span class="ident">access_flags</span>.<span class="ident">contains</span>(<span class="kw-2">&amp;</span><span class="ident">Enum</span>) { <span class="ident">flags</span> <span class="op">^=</span> <span class="ident">ENUM_FLAG</span>; }
        <span class="kw">if</span> <span class="ident">access_flags</span>.<span class="ident">contains</span>(<span class="kw-2">&amp;</span><span class="ident">Module</span>) { <span class="ident">flags</span> <span class="op">^=</span> <span class="ident">MODULE_FLAG</span>; }

        <span class="ident">flags</span>
    }
}
</pre></div>
</section><section id="search" class="content hidden"></section><section class="footer"></section><aside id="help" class="hidden"><div><h1 class="hidden">Help</h1><div class="shortcuts"><h2>Keyboard Shortcuts</h2><dl><dt><kbd>?</kbd></dt><dd>Show this help dialog</dd><dt><kbd>S</kbd></dt><dd>Focus the search field</dd><dt><kbd>↑</kbd></dt><dd>Move up in search results</dd><dt><kbd>↓</kbd></dt><dd>Move down in search results</dd><dt><kbd>↹</kbd></dt><dd>Switch tab</dd><dt><kbd>&#9166;</kbd></dt><dd>Go to active search result</dd><dt><kbd>+</kbd></dt><dd>Expand all sections</dd><dt><kbd>-</kbd></dt><dd>Collapse all sections</dd></dl></div><div class="infos"><h2>Search Tricks</h2><p>Prefix searches with a type followed by a colon (e.g., <code>fn:</code>) to restrict the search to a given type.</p><p>Accepted types are: <code>fn</code>, <code>mod</code>, <code>struct</code>, <code>enum</code>, <code>trait</code>, <code>type</code>, <code>macro</code>, and <code>const</code>.</p><p>Search functions by type signature (e.g., <code>vec -> usize</code> or <code>* -> vec</code>)</p><p>Search multiple things at once by splitting your query with comma (e.g., <code>str,u8</code> or <code>String,struct:Vec,test</code>)</p></div></div></aside><script>window.rootPath = "../../";window.currentCrate = "jvm_class_file_parser";</script><script src="../../aliases.js"></script><script src="../../main.js"></script><script src="../../source-script.js"></script><script src="../../source-files.js"></script><script defer src="../../search-index.js"></script></body></html>