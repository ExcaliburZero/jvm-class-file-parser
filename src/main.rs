extern crate jvm_class_file_parser;

use std::env;
use std::fs;
use std::fs::File;
use std::io;
use std::ops::Deref;
use std::path::PathBuf;

use jvm_class_file_parser::{
    Bytecode, ClassFile, ConstantPoolEntry, ExceptionTableEntry, Method
};

const CONSTURCTOR_NAME: &str = "<init>";

fn main() {
    let args: Vec<String> = env::args().collect();

    let filepath = &args[1];

    let mut file = File::open(filepath).unwrap();
    let class_file = ClassFile::from_file(&mut file).unwrap();

    let absolute_filepath = to_absolute_filepath(filepath).unwrap();

    println!("Classfile {}", absolute_filepath.to_str().unwrap());

    let source_file = class_file.get_source_file_name();
    if let Some(source_file) = source_file {
        println!("  Compiled from: \"{}\"", source_file);
    }

    println!("class {}", class_file.get_class_name());

    println!("  minor version: {}", class_file.minor_version);
    println!("  major version: {}", class_file.major_version);

    print_constant_pool(&class_file);

    println!("{}", "{");

    for method in class_file.methods.iter() {
        print_method(&class_file, method);
    }

    println!("{}", "}");

    if let Some(source_file) = source_file {
        println!("SourceFile: \"{}\"", source_file);
    }

    //println!("{:#?}", class_file);
}

fn to_absolute_filepath(filepath: &str) -> io::Result<PathBuf> {
    let path = PathBuf::from(filepath);

    fs::canonicalize(path)
}

fn print_constant_pool(class_file: &ClassFile) {
    println!("Constant pool:");

    for (i, constant) in class_file.constant_pool.iter().enumerate() {
        // Account for 1 indexing
        let i = i + 1;

        println!(
            "{:>5} = {}",
            format!("#{}", i),
            format_constant_pool_entry(class_file, constant)
        );
    }
}

fn format_constant_pool_entry(
        class_file: &ClassFile, constant: &Box<ConstantPoolEntry>
    ) -> String {
    use ConstantPoolEntry::*;

    match constant.deref() {
        &ConstantUtf8 { ref string } => {
            format!(
                "{:<20}{}",
                "Utf8",
                string
            )
        },
        &ConstantClass { name_index } => {
            format!(
                "{:<20}{:<16}// {}",
                "Class",
                format!("#{}", name_index),
                class_file.get_constant_utf8(name_index as usize)
            )
        },
        &ConstantFieldref { class_index, name_and_type_index } => {
            format!(
                "{:<20}{:<16}// {}",
                "Fieldref",
                format!("#{}.#{}", class_index, name_and_type_index),
                format!(
                    "{}.{}",
                    class_file.get_constant_class_str(class_index as usize),
                    class_file.get_constant_name_and_type_str(
                        name_and_type_index as usize
                    ),
                )
            )
        },
        &ConstantMethodref { class_index, name_and_type_index } => {
            format!(
                "{:<20}{:<16}// {}",
                "Methodref",
                format!("#{}.#{}", class_index, name_and_type_index),
                format!(
                    "{}.{}",
                    class_file.get_constant_class_str(class_index as usize),
                    class_file.get_constant_name_and_type_str(
                        name_and_type_index as usize
                    ),
                )
            )
        },
        &ConstantNameAndType { name_index, descriptor_index } => {
            format!(
                "{:<20}{:<16}// {}",
                "NameAndType",
                format!("#{}:#{}", name_index, descriptor_index),
                format!(
                    "\"{}\":{}",
                    class_file.get_constant_utf8(name_index as usize),
                    class_file.get_constant_utf8(descriptor_index as usize),
                )
            )
        },
        _ => panic!(),
    }
}

fn print_method(class_file: &ClassFile, method: &Method) {
    let method_name = class_file.get_constant_utf8(method.name_index as usize);

    println!(
        "  {}();",
        if method_name == CONSTURCTOR_NAME { class_file.get_class_name() }
            else { method_name }
    );

    println!(
        "    descriptor: {}",
        class_file.get_constant_utf8(method.descriptor_index as usize)
    );

    println!(
        "    flags: TODO",
    );

    let code = method.get_code(class_file).unwrap();

    println!("    Code:");
    println!(
        "      stack={}, locals={}, args_size={}",
        code.max_stack,
        code.max_locals,
        "TODO"
    );

    print_bytecode(class_file, &code.code);

    if code.exception_table.len() > 0 {
        print_exception_table(class_file, &code.exception_table);
    }
}

fn print_bytecode(class_file: &ClassFile, code: &Vec<(usize, Bytecode)>) {
    for (i, bytecode) in code {
        print!(
            "        {:>3}: {:35}",
            i,
            bytecode.to_string()
        );

        // TODO: show constants to the side

        println!();
    }
}

fn print_exception_table(
        class_file: &ClassFile, exception_table: &Vec<ExceptionTableEntry>
    ) {
    println!("      Exception table:");
    println!("         from    to  target type");

    for entry in exception_table.iter() {
        println!(
            "         {:5} {:5} {:5}   Class {}",
            entry.start_pc,
            entry.end_pc,
            entry.handler_pc,
            class_file.get_constant_class_str(entry.catch_type as usize),
        );
    }
}
