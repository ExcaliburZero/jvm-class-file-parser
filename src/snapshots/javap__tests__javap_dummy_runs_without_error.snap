---
source: src/main.rs
expression: "javap(\"classes/Dummy.class\", true)"
---
"Classfile /home/jbalint/sw/rust-sw/jvm-class-file-parser/classes/Dummy.class\n  Compiled from: \"Dummy.java\"\nclass Dummy\n  minor version: 0\n  major version: 55\n  flags: ACC_PUBLIC, ACC_SUPER\nConstant pool:\n   #1 = Methodref           #3.#10          // java/lang/Object.\"<init>\":()V\n   #2 = Class               #11             // Dummy\n   #3 = Class               #12             // java/lang/Object\n   #4 = Utf8                <init>\n   #5 = Utf8                ()V\n   #6 = Utf8                Code\n   #7 = Utf8                LineNumberTable\n   #8 = Utf8                SourceFile\n   #9 = Utf8                Dummy.java\n  #10 = NameAndType         #4:#5           // \"<init>\":()V\n  #11 = Utf8                Dummy\n  #12 = Utf8                java/lang/Object\nAttributes:\n  SourceFile = \"Dummy.java\"\n{\n  Dummy();\n    descriptor: ()V\n    flags: TODO\n    Code:      stack=1, locals=1, args_size=TODO          0: aload_0                            \n          1: invokespecial #1                   \n          4: return                             \n}\nSourceFile: \"Dummy.java\"\n"
